File: ecmascript/beer-song/beer-song.js
let verse = (num) => {
  if (num > 0) {
    let n = num
    let p = n === 1 ? '' : 's'
    let n1 = n === 1 ? 'no more' : num - 1
    let p1 = n1 === 1 ? '' : 's'
    let c = n === 1 ? 'it' : 'one'
    return `${n} bottle${p} of beer on the wall, ${n} bottle${p} of beer.
Take ${c} down and pass it around, ${n1} bottle${p1} of beer on the wall.
`
  } else if (num === 0) {
    return `No more bottles of beer on the wall, no more bottles of beer.
Go to the store and buy some more, 99 bottles of beer on the wall.
`
  }
}

let sing = (startNum = 99, endNum = 0) => {
  let verses = ''
  for (let i = startNum; i >= endNum; i--) {
    verses += verse(i)
    if (i !== endNum) verses += '\n'
  }
  return verses
}
export default {
  verse: verse,
  sing: sing
}

File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(n) do
    cond do
      n == 1 ->
        "No more bottles of beer on the wall, no more bottles of beer.\n" <>
          "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
      n == 2 ->
        Integer.to_string(n-1) <> " bottle of beer on the wall, " <> Integer.to_string(n-1) <>
          " bottle of beer.\nTake it down and pass it around, "<> "no more"<>
          " bottles of beer on the wall.\n"
      n == 3 ->
        Integer.to_string(n-1) <> " bottles of beer on the wall, " <> Integer.to_string(n-1) <>
          " bottles of beer.\nTake one down and pass it around, "<> Integer.to_string(n-2) <>
          " bottle of beer on the wall.\n"
      true ->
        Integer.to_string(n-1) <> " bottles of beer on the wall, " <> Integer.to_string(n-1) <>
          " bottles of beer.\nTake one down and pass it around, "<> Integer.to_string(n-2) <>
          " bottles of beer on the wall.\n"
    end
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range) do
    returnString = Enum.reduce(range, "", fn (i, acc) -> acc <> verse(i) <> "\n" end)
    String.slice(returnString, 0..-2)
  end

  def lyrics() do
    lyrics(100..1)
  end
end
